### PR TITLE
Clippy Book Chapter Updates Reborn: Defining Lints

### PR DESCRIPTION
**Pull request by blyxyas**
Revival of #9659, I've made a few changes (mainly ones from reviews from #9426, like removing mid-section storytelling) and some re-writes. The goal of the project would be to slowly re-write the Clippy book (only chapters that need it) to more up-to-date style.

## Notes

- Things like `git status` commands have changed, we need to make sure that they're correct.
- As this is a team-wide effort, I would please ask anyone and everyone to read it and give your opinion.

---

changelog: Add a new "Defining lints" chapter to the book

r? @flip1995

### CODE REVIEW

**Comment by blyxyas** [comment]
I consider documentation very important and a team-wide effort, could you please read it and give your opinions (On a review or a one-off comment)?
@rustbot ping clippy

**Comment by blyxyas** [comment]
Weird

cc @rust-lang/clippy

File: book/src/development/defining_lints.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+$ cargo dev new_lint --name=foo_functions --pass=late --category=pedantic
+```
+
+There are two things to note here:
+
+1. We set `--pass=late` in this command to do a late lint pass. The alternative
+is an `early` lint pass. We will discuss this difference in a later chapter.
+<!-- FIXME: Link that "later chapter" when lint_passes.md is merged -->
+2. If not provided, the `category` of this new lint will default to `nursery`.
+See Clippy's [lint types](../lints.md) for more information on categories.
```
Comment (llogiq) [inline book/src/development/defining_lints.md:44]: Here we have `lint types` linking to the upper chapter, and elsewhere there's a Lint Types heading, which makes me think the term might be overloaded. If so, could we find a more apt term here?

**Review by llogiq — APPROVED** [review]
I'm not completely clear about the planned extent of this chapter, so I'm not sure if there's anything missing, but what I see looks good.

Perhaps adding some copy from the created files to show what should be changed (notably the lint declaration itself, e.g. the version annotation often trips up newbies) might help, but also take up some space, and I'm not sure if you want to tackle that in a "documenting lints" chapter.

File: book/src/development/defining_lints.md (hunk 1 above)
Comment (blyxyas) [reply book/src/development/defining_lints.md:44]: Maybe changing that to link to the later section "Lint types", and adding a note in "Lint types" about how there's more information in `../lints.md` could clear the confusion?

Comment (blyxyas) [reply book/src/development/defining_lints.md]: I think that the latest commit would solve the issue.

File: book/src/development/defining_lints.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+When a lint is manually declared, we might need to register the lint pass
+manually in the `register_plugins` function in `clippy_lints/src/lib.rs`:
+
+```rust
+store.register_late_pass(|| Box::new(foo_functions::FooFunctions));
+```
+
+As you might have guessed, where there's something late, there is something early:
+in Clippy there is a `register_early_pass` method as well.
+More on early vs. late passes in a later chapter.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:132]: I would replace all of this with just a one line code snippet.

````suggestion
Or for early lint passes:

```rust
store.register_early_pass(|| Box::new(foo_functions::FooFunctions));
```
````

File: book/src/development/defining_lints.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+the lint will be automatically registered and there is nothing more to do.
+
+However, sometimes we might want to declare a new lint by hand.
+In this case, we'd use `cargo dev update_lints` command afterwards.
+
+When a lint is manually declared, we might need to register the lint pass
+manually in the `register_plugins` function in `clippy_lints/src/lib.rs`:
+
+```rust
+store.register_late_pass(|| Box::new(foo_functions::FooFunctions));
```
Comment (flip1995) [inline book/src/development/defining_lints.md:127]:
```suggestion
store.register_late_pass(|_| Box::new(foo_functions::FooFunctions));
```

File: book/src/development/defining_lints.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+### Specific Type
+
+> **Note**: Lint types are listed in the ["Lint types"](#lint-types) section
+
+If you believe that this new lint belongs to a specific type of lints,
+you can run `cargo dev new_lint` with a `--type` option.
+
+Since our `foo_functions` lint is related to function calls, one could
+argue that we should put it into a group of lints that detect some behaviors
+of functions, we can put it in the `functions` group.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:76]: Besides other things, this can be removed if using `generic_lint_name` instead for example commands.

File: book/src/development/defining_lints.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+In our case, the path will be `clippy_lints/src/functions/foo_functions.rs`.
+
+Notice how this command has a `--type` flag instead of `--pass`. Unlike a standalone
+definition, this lint won't be registered in the traditional sense. Instead, you will
+call your lint from within the type's lint pass, found in `clippy_lints/src/{type}/mod.rs`.
+
+A _type_ is just the name of a directory in `clippy_lints/src`, like `functions` in
+the example command. Clippy groups together some lints that share common behaviors,
+so if your lint falls into one, it would be best to add it to that type.
+Read more about [lint types](#lint-types) below.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:95]: This can also be completely removed IMO, if the `Lint Types` section gets moved to the top of the document. And then give it more of a "Where to put a new lint?" feeling to it.

**Review by flip1995 — COMMENTED** [review]
I would like this documentation get way slimmer. This still seems like a lot of stuff is copied over from `adding_lints.md`. This should be a short explanation on how to create lints. The example for `foo_functions` should be almost completely removed. I would either replace it with some generic `lint_name` in the example commands or keep the name, but then don't talk about what it does at all. This should go into a full-tutorial documentation, like what `adding_lints` is today.

I would also move the `Lint Types` up.

I'm unsure if `Lint Name` is placed well in this chapter. I'm inclined to completely remove it or slim it down to just one sentence that points to `lint naming guidelines`.

And finally, as llogiq pointed out, I would at least copy the `define_clippy_lints` macro that is generated into this documentation and explain each part and how it should be updated (documentation + version) before submitting the lint.

---

Basically what I'd like this document to become is a quick overview/collection of commands that could be run to generate new lints plus explanation what is generated, if it has to be generated manually.

**Comment by blyxyas** [comment]
The latest commits should fix all issues

File: book/src/development/defining_lints.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+The first step in the journey of a new lint is the definition
+and registration of the lint in Clippy's codebase.
+We can use the Clippy dev tools to handle this step since setting up the 
+lint involves some boilerplate code.
+
+#### Lint types
+
+A lint type is the category of items and expressions in which your lint focuses on.
+
+As of the writing of this documentation update, there are 12 groups (a.k.a. _types_)
```
Comment (flip1995) [inline book/src/development/defining_lints.md:12]:
```suggestion
As of the writing of this documentation update, there are 12 _types_
```

I would just consistently call it lint types to avoid all confusion with lint groups.

File: book/src/development/defining_lints.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+- `misc_early`
+- `operators`
+- `transmute`
+- `types`
+- `unit_types`
+- `utils / internal` (Clippy internal lints)
+
+These types group together lints that share some common behaviors.
+For instance, `functions` groups together lints
+that deal with some aspects of function calls in Rust.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:30]: Not blocking: We might want to add a one-line description for each type in the list above.

Question: Are function lints really about function calls, rather than function signatures/declarations? :thinking:

File: book/src/development/defining_lints.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+```sh
+$ cargo dev new_lint --name=lint_name --pass=late --category=pedantic
+```
+
+There are two things to note here:
+
+1. We set `--pass=late` in this command to do a late lint pass. The alternative
+is an `early` lint pass. We will discuss this difference in a later chapter.
+<!-- FIXME: Link that "later chapter" when lint_passes.md is merged -->
+2. If not provided, the `category` of this new lint will default to `nursery`.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:65]:
```suggestion
1. `--pass`: We set `--pass=late` in this command to do a late lint pass. The alternative
is an `early` lint pass. We will discuss this difference in a later chapter.
<!-- FIXME: Link that "later chapter" when lint_passes.md is merged -->
2. `--category`: If not provided, the `category` of this new lint will default to `nursery`.
```

File: book/src/development/defining_lints.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+In our case, the path will be `clippy_lints/src/functions/foo_functions.rs`.
+
+Notice how this command has a `--type` flag instead of `--pass`. Unlike a standalone
+definition, this lint won't be registered in the traditional sense. Instead, you will
+call your lint from within the type's lint pass, found in `clippy_lints/src/{type}/mod.rs`.
+
+A _type_ is just the name of a directory in `clippy_lints/src`, like `functions` in
+the example command. Clippy groups together some lints that share common behaviors,
+so if your lint falls into one, it would be best to add it to that type.
+Read more about [lint types](#lint-types) below.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:118]:
```suggestion
```

The note on top of this section already covers this.

File: book/src/development/defining_lints.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	/// 
+	/// Triggers when detects...
+	/// 
+    /// ### Why is this bad?
+	/// 
+	/// // Describe why this pattern would be bad
+	/// 
+	/// It can lead to...
+    ///
+    /// ### Example
```
Comment (flip1995) [inline book/src/development/defining_lints.md:160]:
```suggestion
    /// ### What it does
    ///
    /// // Describe here what does the lint do.
    /// 
    /// Triggers when detects...
    /// 
    /// ### Why is this bad?
    /// 
    /// // Describe why this pattern would be bad
    /// 
    /// It can lead to...
    ///
    /// ### Example
```

File: book/src/development/defining_lints.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+When a lint is manually declared, we might need to register the lint pass
+manually in the `register_plugins` function in `clippy_lints/src/lib.rs`:
+
+```rust
+store.register_late_pass(|| Box::new(foo_functions::FooFunctions));
+```
+
+As you might have guessed, where there's something late, there is something early:
+in Clippy there is a `register_early_pass` method as well.
+More on early vs. late passes in a later chapter.
```
Comment (flip1995) [inline book/src/development/defining_lints.md:192]:
```suggestion
More on early vs. late passes in a later chapter.
<!-- FIXME: Link that "later chapter" when lint_passes.md is merged -->
```

File: book/src/development/defining_lints.md (hunk 7 above)
Comment (flip1995) [reply book/src/development/defining_lints.md]: Yes it is mostly about declarations/signatures. I changed the wording a bit here.

**Comment by flip1995** [comment]
@bors r+

I also rebased the branch on top of latest master. Let's get those PRs in :shipit:
